spring-integration-mongodb/src/main/java/org/springframework/integration/mongodb/metadata/MongoDbMetadataStore.java

```diff
@@ -0,0 +1,197 @@
+		this(template, null);
+	}
+
+	/**
+	 * Configures the MongoDbMetadataStore by provided {@link MongoDbFactory} and
+	 * default collection name - {@link #DEFAULT_COLLECTION_NAME}.
+	 * 
+	 * @param factory The mongodb factory
+	 */
+	public MongoDbMetadataStore(MongoDbFactory factory) {
+		this(factory, null);
+	}
+	
+	/**
+	 * Configures the MongoDbMetadataStore by provided {@link MongoDbFactory} and
+	 * collection name
+	 * 
+	 * @param collectionName the collection name where it persists the data
+	 * @param factory MongoDbFactory
+	 */
+	public MongoDbMetadataStore(MongoDbFactory factory, String collectionName) {
+		template = new MongoTemplate(factory);
+		this.collectionName = (StringUtils.hasText(collectionName)) ? collectionName
+				: DEFAULT_COLLECTION_NAME;
+	}
+	
+	
+	/**
+	 * Method put.
+	 * @param key String
+	 * @param value String
+	 * @see org.springframework.integration.metadata.MetadataStore#put(String, String)
+	 */
+	@Override
+	public void put(String key, String value) {
+		Assert.notNull(key, "'key' must not be null.");
+		Assert.notNull(value, "'value' must not be null.");
+		Map<String,String> fileInfo = new HashMap();
+		fileInfo.put(ID_FIELD, key);
+		fileInfo.put(VALUE, value);
+		template.save(fileInfo, collectionName);
+	}
+
+	/**
+	 * Method get.
+	 * @param key String
+	 * @return String
+	 * @see org.springframework.integration.metadata.MetadataStore#get(String)
+	 */
+	@Override
+	public String get(String key) {
+		Assert.notNull(key, "'key' must not be null.");
+		Query query = new Query(Criteria.where(ID_FIELD).is(key));
+		@SuppressWarnings("unchecked")
+		Map<String,String> result = template.findOne(query, Map.class, collectionName);
+		return result==null?null:result.get(VALUE);
+		
+	}
+
+	/**
+	 * Method remove.
+	 * @param key String
+	 * @return String
+	 * @see org.springframework.integration.metadata.MetadataStore#remove(String)
+	 */
+	@Override
+	public String remove(String key) {
+		Assert.notNull(key, "'key' must not be null.");
+		Query query = new Query(Criteria.where(ID_FIELD).is(key));
+		@SuppressWarnings("unchecked")
+		Map<String,String> result = template.findAndRemove(query, Map.class,collectionName);
+		return result==null?null:result.get(VALUE);
+	}
+
+	/**
+	 * Method putIfAbsent.
+	 * @param key String
+	 * @param value String
+	 * @return String
+	 * @see org.springframework.integration.metadata.ConcurrentMetadataStore#putIfAbsent(String, String)
+	 */
+	@Override
+	public String putIfAbsent(String key, String value) {
+		Assert.notNull(key, "'key' must not be null.");
+		Assert.notNull(value, "'value' must not be null.");
+
+		String result = null;
+		Map<String,String> fileInfo = new HashMap();
+		fileInfo.put(ID_FIELD, key);
+		fileInfo.put(VALUE, value);
+		try {
+			template.insert(fileInfo, collectionName);
+		} catch (DuplicateKeyException e) {
+			if(logger.isDebugEnabled()){
+				logger.debug("DUPLICATE KEY VIOLOATION",e);
```

Comment:
Typo. However I don't think that we should say this `debug` at all, because it is normal behaviour for `putIfAbsent`  contract.
From other side we talked with @christophstrobl about this operation and found that it can be achieved only with the `raw MongoDB script execution` to be sure in the atomicity.
This feature is available since the last Spring Data Fowler and SD-Mongo-1.7.0.
Spring Integration is already based on the last SD, but you should rebase your branch to the last `master`.

How script looks we'll show you later: after fixing for all other concerns to this PR.